Review PR

Mikailus opened the pull request.

apyd committed:
Project structure

apyd committed:
HTML project structure + color and font variables

apyd committed:
Added buttons, missing classes to elements

apyd committed:
Added mixins and layout stylings

apyd committed:
Removed rows within board, added basic styings for board

apyd committed:
Button hover effect and equal spacing between cells (inner and outer)

apyd committed:
Added rules popup, basic model/view/controller classes

apyd committed:
Tiles styling

apyd committed:
Modules export, validating key pressed in controller

apyd committed:
Creating and adding a tile in view

apyd committed:
Generating coordinates for new tile

apyd committed:
Updated coordinates (to reflect array coordinates)

apyd committed:
Adding tile on arrow key press and counting tiles on board

apyd committed:
Moving tiles on right arrow key press in model

apyd committed:
Model  - sort tiles to left on left arrow key press

apyd committed:
Fix for sorting arrays on left arrow key press

apyd committed:
Close popup on 'ESC' key press

apyd committed:
Move tiles on up/down arrow key press

apyd committed:
Merge tiles and move on up/down arrow key press

apyd committed:
Refactor - tiles movement

apyd committed:
Merge tiles

apyd committed:
Update score on tiles merge

apyd committed:
Store bestResult in localstorage and updating score/bestScore values on arrow key press

apyd committed:
Default export for model/view/controller and creating localStorage bestScore item if doesn't exist on init

apyd committed:
Keeping bestScore in localStorage, updating and showing score and bestScore on tiles merge

apyd committed:
Game blocked before 'start game' button pushed, two tiles added on game init

apyd committed:
Added favicon

apyd committed:
Tiles added directly to board  /  using board background image to imitate cells

apyd committed:
Refactor app.js and view.js / adding tiles directly to board (positioned absolutely)

apyd committed:
Clear board on start game

apyd committed:
Count tiles on board

apyd committed:
Each tile has own id and stored as object, each tile on view has own id data attribute as well

apyd committed:
Scoreboard left margin change

apyd committed:
Fix to merge tiles while they are objects

apyd committed:
Sending array with updated tiles position to view

apyd committed:
Using older, widely supported way of returning flat array

apyd committed:
Moving tiles on arrow key press, storing id from merged tile in retained tile

apyd committed:
Merge tiles in view

apyd committed:
Clear previous score when game is started, load bestScore on DOMContentLoaded

apyd committed:
Animate on tile add, move and tiles merge

apyd committed:
Remove animation/transition class after timeout

apyd committed:
Increased animation/transition duration and fixed footer width on smaller screen resolution

apyd committed:
Repeating sort to avoid empty space on merge tiles and reducing animation duration

apyd committed:
Added buttons with arrows for mobile devices

apyd committed:
Entry screen with game type selection

apyd committed:
Creating board in model based on selected game type

apyd committed:
Styling improvements

apyd committed:
Use appropriate board size and tile size on selected game type

apyd committed:
Different tile dimension based on selected game type

apyd committed:
Creating multidimensional array based on game type selection

apyd committed:
Fix for retained tile position

apyd committed:
Storing separately best result for 3x3 and 4x4 board and showing on game init

apyd committed:
Animate board on game type selection

apyd committed:
Added rules button on entry screen and reduced opacity on unchecked game type

apyd committed:
Move tiles on swipe (touch screens)

apyd committed:
Handling all buttons, added left arrow button

apyd committed:
Left arrow allows to go back to entry screen and change game type

apyd committed:
Detect end of game

apyd committed:
Fix for end game indicaton

apyd committed:
Replaced game type 3x3 with 5x5

apyd committed:
Show popup with details on game end

apyd committed:
Clear score on each game start

apyd committed:
Get best score value on game init to avoid missing item in local storage

apyd committed:
Remove unused icons/images

apyd committed:
Styling for end game popup

apyd committed:
RWD improvements

apyd committed:
RWD for entry screen

apyd committed:
RWD for entry screen

apyd committed:
Images replacement (better sharpness) and tiles position fix

apyd committed:
Recalculate tiles position on screen resize

apyd committed:
Recalculate tiles position on screen resize, board images replacement

apyd committed:
RWD for mobile screen

apyd committed:
Disable refresh on swipe down on mobile browser

apyd committed:
Detect if user move cause changes on board, if not skip the move

apyd committed:
Detect swipe direction only if distance between points reached minimum

apyd committed:
Reduce tile font size if it's value exceeds 4 digits

apyd committed:
Styling improvements

apyd committed:
CSS refactor

apyd committed:
CSS refactor

apyd committed:
CSS refactor

apyd committed:
Fix for showing appropriately converted time elapsed and refactor toggle popup function

apyd committed:
View refactor

apyd committed:
RWD fix and removed unecessary wrapper for button

apyd committed:
Text change in header

apyd committed:
Added stylings for higher tile values, changes to animations

apyd committed:
Reduced timeout length and text change in footer

apyd committed:
Readme.md update

apyd committed:
Changes to animations

apyd committed:
Merge branch 'main' of https://github.com/apyd/2048-game into main

apyd committed:
Readme.md update

apyd committed:
CSS refactor

apyd committed:
Merge branch 'main' of https://github.com/apyd/2048-game into main

Mikailus commented on assets/css/_mixins.scss:
```diff
@@ -0,0 +1,32 @@
+@mixin positionFlex($direction, $align) {
```
Ten mixin ogranicza Cię do pozycjonowania horyzontalnie do środka - nie lepiej pozwolić też na ustawianie `justify-content`?

apyd replied:
Masz racje. W tym projekcie nigdy nie potrzebowałem innej wartości dla justify-content ale nazwa mixina nie wskazuje na takie ustawienie więc zaktualizowałem zgodnie z uwagą.

Mikailus commented on assets/css/_mixins.scss:
```diff
@@ -0,0 +1,32 @@
+    animation-duration: $duration;
+}
+
+@mixin tablet {
```
Odnośnie tego typu mixinów rozmawialiśmy wczoraj

Mikailus commented on assets/css/_variables.scss:
```diff
@@ -0,0 +1,26 @@
+$selective-yellow: #F7B801;
+$tangerine: #F18701;
+$orange-pantone: #F35B04;
+$tile2--bg: #FFE55E;
```
Z tych wszystkich zmiennych `$tile-...` można by było zrobić mapę `tiles-bg` i wyciągać z niej poszczególne kolory - bo lepiej będą zgrupowane

Mikailus commented on assets/css/_variables.scss:
```diff
@@ -0,0 +1,26 @@
+$tile16384--bg: #65070C;
+$light-gray: #F2F2F2;
+$white: #FFFFFF;
+$screen-s-max: 480px;
```
Analogicznie jak dla kolorów kafelków - breakpointy zazwyczaj jako mapy się tworzy

Mikailus commented on assets/css/style.css:
```diff
@@ -0,0 +1,742 @@
+          box-sizing: border-box;
+}
+
+.html {
```
Nie musisz na siłę tworzyć klasy dla znacznika `html` - on domyślnie jest tylko jeden na stronie, więc możesz go spokojnie po znaczniku ostylować

Mikailus commented on assets/css/style.css:
```diff
@@ -0,0 +1,742 @@
+  height: 100%;
+}
+
+.body {
```
Analogicznie jak dla `.html`

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+}
+
+.body {
+    font-family: $inter;
```
Zwróć uwagę na kolejność właściwości, jest to bardzo pomocne potem: https://css-tricks.com/poll-results-how-do-you-order-your-css-properties/

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+@import './variables';
+@import './mixins';
```
Poczytaj o `@use`

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+    }
+}
+
+.btn {
```
Wiesz co masz z tym zrobić :D

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+        font-size: 1.2rem;
+    }
+
+    &.btn--primary {
```
1. Można zrobić: `&.&--primary`
2. Potrzebujemy zagnieżdżać w nim, nie możemy utrzymać płaskiej struktury CSS?

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+        margin: 5px;
+    }
+
+    &.label--4 {
```
`.label--4` nie jest semantyczną nazwą dla modyfikatora

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+        }
+    }
+
+    &.label--5 {
```
Analogicznie jak dla `label--5`

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+    @include positionFlex(column, center);
+    width: 100vw;
+
+    .header > .header__text {
```
Dlaczego takie zagnieżdżenie? Z BEM jasno wynika, że to element nagłówka

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+        justify-content: space-evenly;
+        width: 90vw;
+
+        .radio:not(:checked) + label img {
```
1. Nie utrzymujesz płaskiej struktury .row > .radio + label > img
2. Mieszasz klasy i znaczniki
3. Jeżeli `.radio` ma zachowywać się inaczej wewnątrz `.row`, to `radio` powinno mieć odpowiedni modyfikator - `.radio` jest niezależnym komponentem, więc nie powinieneś sterować jego wyglądem z poziomu innego elementu

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+            opacity: .6;
+        }
+
+        .radio__img {
```
```
.radio {
  &__img {}
  &__text {}
  }
```

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+            margin: 5px 0 0 0;
+        }
+
+        .btn {
```
Jeżeli `.btn` ma zachowywać się inaczej wewnątrz `.row`, to `btn` powinno mieć odpowiedni modyfikator - `.btn` jest niezależnym komponentem, więc nie powinieneś sterować jego wyglądem z poziomu innego elementu

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+    background-color: $black-7-opacity;
+    z-index: 1000;
+
+    .popup {
```
Czy popup może pojawić się bez overlay?

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+                margin: 0;
+            }
+
+            .btn--close {
```
`.btn--close` powinien być tam, gdzie jest `.btn`

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+            position: relative;
+            padding-top: 16px;
+
+            .header__text {
```
Czy przypadkiem tej klasy nie używasz już dla nagłówka strony?

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+            width: 80vw;
+        }
+
+        .popup__header {
```
```
.popup {
  &__header {}
}
```

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+                    background-color: $light-gray;
+                }
+
+                .icon--close {
```
`.icon--close` powinien być tam, gdzie jest `.icon`

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+            }
+    }
+
+            .popup__body {
```
```
.popup {
  &__body{}
}
```

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+                flex: 1;
+                font-size: vmin(4vw, 4vh);
+
+                .text__label {
```
`.text__label` powinien być tam, gdzie jest `.text`

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+    }
+
+.main {
+    display: flex;
```
Czemu nie używasz tutaj mixina?

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+    width: 100%;
+    overflow: hidden;
+
+    .header__text {
```
Powinno być tam, gdzie `header`

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+                    font-size: 1rem; 
+                }
+
+                .scoreboard__value {
```
Piąty poziom zagnieżdżenia - stanowczo za głęboko. Czy `.scoreboad` nie jest niezależnym komponentem strony? Imo jest, więc nie powinien być zagnieżdżony nigdzie

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+        text-align: center;
+    }
+
+    .row {
```
Z nazwy klasy wnioskuję, że `.row` traktujesz jako komponent - więc nie powinien być zagnieżdżony w klasie `.main`

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+        height: 300px;
+    }
+
+    &#t4 {
```
Włąściwości wewnątrz tego bloku i bloku `t5` są niemal takie same - może zrób mixin, który będzie dostawał jako argument urla do obrazka i zwróci
```
        background: url('../img/5x5desktop.png') no-repeat no-repeat;
        @include animation(blur, 400ms) {
            0% { -webkit-filter: blur(20px);}
            100% { -webkit-filter: blur(0px);}
        }
        @include mobile {
            background: url('../img/5x5mobile.png') no-repeat no-repeat;
        }
 ```
 Z podmianą odpowiednich URLi

apyd replied:
Udało mi się zredukować powtarzalne właściwości i umiejscowiłem je w board
Po zmianie:

```
    &--4x4 {
        background-image: url('../img/4x4desktop.png');
        @include mixins.respond-to('mobile') {
            background-image: url('../img/4x4mobile.png');
        }
    }
    
    &--5x5 {
        background-image: url('../img/5x5desktop.png');
        @include mixins.respond-to('mobile') {
            background-image: url('../img/5x5mobile.png');
        }
    }
```
patrząc z perspektywy używalności mixina w takim przypadku wydaje mi się, że byłby to przerost formy nad treścią(?)
Bo tylko podmieniam background image a wszystkie pozostałe właściwości i animacja już definiuje w board.

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+            }
+        }
+
+        &.tile--2 {
```
Poczytaj o https://sass-lang.com/documentation/at-rules/control/each - znacznie skróci Tobie linijki poniżej

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+        }
+    }
+
+    .tile {
```
Czy `.tile` musi być zagnieżdżony w `.main`? Nie może być traktowany niezależnie?

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+        }
+
+
+
```
Pusta zbędna linijka

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+            }
+        }
+        
+            .tile__value {
```
Wcięcie jest niepoprawne chyba

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+    }
+}
+
+.hidden {
```
Tę klasę bym wyciągnął do jakiegoś pliku utilsowego, bo to dość generyczna klasa

Mikailus commented on assets/css/style.scss:
```diff
@@ -0,0 +1,463 @@
+        color: $selective-yellow;
+        font-weight: bold;
+    }
+}
```
W kilku miejscach, jak masz różne komponenty, to warto potworzyć może plik per komponent - ja bym się zastanowił nad takim podejściem

Mikailus commented on assets/js/controller.js:
```diff
@@ -0,0 +1,93 @@
+export default class Controller {
+    constructor() {
+        this.gameStatus = 0;
+        this.tilesOnBoard;
```
Jak nie inicjalizujesz tych zmiennych w konstruktorze, to możesz je podać poza konstruktorem:
```
class MyClass {
  tilesOnBoard;
  gameType;
}
```

Mikailus commented on assets/js/controller.js:
```diff
@@ -0,0 +1,93 @@
+        this.initY;
+        this.initX;
+        this.startGameTime;
+        this.EndGameTime;
```
`endGameTime`

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    initGame(selectedType) {
+            this.gameType = +selectedType;
```
Wcięcia

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+
+    initGame(selectedType) {
+            this.gameType = +selectedType;
+            this.board = Array(this.gameType).fill().map(() => Array(this.gameType).fill(''));
```
```
createBoard() {
  return Array(this.gameType).fill().map(() => Array(this.gameType).fill(''))
}
```

```this.board = this.createBoard();```

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+
+export default class Model {
+    constructor() {
+        this.nextId = 0;
```
Co przechowuje `nextId`? Po nazwie nie za bardzo wiadomo

apyd replied:
Zmieniłem nazwę na nextAvailableTileId - zmienna przechowuje kolejne dostępne id dla nowej kafelki do użycia

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    clearBoard() {
+        this.board = Array(this.gameType).fill('').map(() => Array(this.gameType));
```
Ta linijka jest łudząco podobna do tej zrefaktorowanej w komentarzu powyżej

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+        this.canAddTile = true;
+    }
+
+    initGame(selectedType) {
```
`initializeGame` - nie jest to aż tak długa nazwa, by warto było ją jeszcze skracać

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    generateCoordinates() {
+        let x = Math.floor((Math.random() * this.gameType));
```
```
function getRandomCoordinate(maxBoundary) {
  return Math.floor((Math.random() * maxBoundary));
}
```

I ta funkcja najlepiej jakby byłą w jakimś pliku o nazwie `utils.js` czy coś takiego

Mikailus replied:
Poza tym `const x`

apyd replied:
Stworzyłem funkcję 

```
function generateRandomCoordinates(maxBoundary) {
    return {
        x: Math.floor((Math.random() * maxBoundary)),
        y: Math.floor((Math.random() * maxBoundary))
    }
}
```

tak aby od razu generować obiekt ze współrzędnymi. Za każdym razem będzie potrzebny 'x' i 'y'. 
W przypadku mojej funkcji potrzebne jest pojedyncze wywołanie. W Twoim przypadku musiałbym ją wykonywać każdorazowo dwukrotnie.
W związku z tym, że współrzędne są losowe i mogą być już zajęte konieczne jest jej ponowne wywołanie aż do momentu otrzymania współrzędnych jeszcze nie użytych.

apyd replied:
@Mikailus w sumie tutaj odwołujesz się do takiego rozwiązania:
https://github.com/apyd/2048-game/pull/1#discussion_r554560475

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+
+    generateCoordinates() {
+        let x = Math.floor((Math.random() * this.gameType));
+        let y = Math.floor((Math.random() * this.gameType));
```
`const y`

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    checkCollision(x, y) {
+        return this.board[x][y] ? false : true;
```
`return !this.board[x][y]`

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+
+    addTile() {
+            if (!this.canAddTile) return;
+            let coordinates = this.generateCoordinates();
```
`const`

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+    addTile() {
+            if (!this.canAddTile) return;
+            let coordinates = this.generateCoordinates();
+            let id = this.nextId;
```
`const`

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+            if (!this.canAddTile) return;
+            let coordinates = this.generateCoordinates();
+            let id = this.nextId;
+            this.board[coordinates.charAt(0)][coordinates.charAt(1)] = {
```
Nie lepiej, by metoda `generateCoordinates` zwracała obiekt o strukturze: `{ x: wartosc_x, y: wartosc_y }`? Wtedu tutaj byłoby `this.board[coordinates.x][coordinates.y]`

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+            let coordinates = this.generateCoordinates();
+            let id = this.nextId;
+            this.board[coordinates.charAt(0)][coordinates.charAt(1)] = {
+                id: id,
```
Możliwe, że wystarczyłoby tutaj samo `id`, zamiast `id: id`

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+                val: 2
+            };
+            this.nextId++;
+            controller.tilesOnBoard++;
```
Dobrą praktyką jest, by klasa sama modyfikowała swoje pola. Czyli w tym przypadku klasa `Controller` mogłaby mieć metodę
```
incrementNumberOfTilesOnBoard() {
  this.tilesOnBoard = ++this.tilesOnBoard;
}
```

I w tym miejscu wywoływałbyś metodę poprzez `controller.incrementNumberOfTilesOnBoard()`

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+            this.nextId++;
+            controller.tilesOnBoard++;
+            return {
+                y: coordinates.charAt(0),
```
Kolejny argument za przekazywaniem obiektu a nie stringa

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+
+    moveTiles(key) {
+        let boardToSort = [...this.board];
+        let sortedBoard = Array(this.gameType).fill().map(() => Array(this.gameType));
```
Znowu łudząco podobny fragment co w `initGame`

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+    moveTiles(key) {
+        let boardToSort = [...this.board];
+        let sortedBoard = Array(this.gameType).fill().map(() => Array(this.gameType));
+        (key === 'ArrowUp' || key === 'ArrowDown') ? boardToSort = this.rotateBoard([...boardToSort]): null;
```
Tutaj chyba bardziej chciałeś coś takiego
```
boardToSort = (key === 'ArrowUp' || key === 'ArrowDown') ? this.rotateBoard([...boardToSort]) : boardToSort;
```

Mikailus replied:
Widzę, że w linijce 73 robisz podobną rzecz, ale na innej tablicy - lepiej wobec tego utworzyć nową metodę
```
jakasMetoda(board) {
  return (key === 'ArrowUp' || key === 'ArrowDown') ? this.rotateBoard([...board]) : board;
}
```

I używać jej

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+        (key === 'ArrowUp' || key === 'ArrowDown') ? boardToSort = this.rotateBoard([...boardToSort]): null;
+        if (key === 'ArrowRight' || key === 'ArrowDown') {
+            for (let [index, subArray] of boardToSort.entries()) {
+                sortedBoard[index] = subArray.filter(el => !el).concat(this.mergeTiles(subArray.filter(el => el), key));
```
Wyjaśnij mi proszę co robi ta linijka

apyd replied:
Prawdopodobnie nieaktualne, metoda wygląda teraz trochę inaczej

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+        if (key === 'ArrowRight' || key === 'ArrowDown') {
+            for (let [index, subArray] of boardToSort.entries()) {
+                sortedBoard[index] = subArray.filter(el => !el).concat(this.mergeTiles(subArray.filter(el => el), key));
+                sortedBoard[index] = [...sortedBoard[index]].filter(el => !el).concat([...sortedBoard[index]].filter(el => el));
```
Analogicznie proszę o wyjaśnienie tutaj

apyd replied:
Prawdopodobnie nieaktualne, metoda wygląda teraz trochę inaczej

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+        let previouslyUpdatedTiles = this.generateArrayWithUpdatedTiles(this.board);
+        this.board = [...sortedBoard];
+        let arrWithUpdatedTiles = this.generateArrayWithUpdatedTiles(sortedBoard);
+        this.checkIfArraysEqual(previouslyUpdatedTiles, arrWithUpdatedTiles) ? this.canAddTile = false : this.canAddTile = true;
```
```
this.canAddTile = !this.checkIfArraysEqual(previouslyUpdatedTiles, arrWithUpdatedTiles)
```

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+        if (typeof arr1 === 'undefined') return false;
+        let previouslyUpdatedTiles = JSON.stringify(arr1);
+        let arrWithUpdatedTiles = JSON.stringify(arr2);
+        let results;
```
`return previouslyUpdatedTiles === arrWithUpdatedTiles`

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+                }
+            }));
+        }
+        let arr = [].concat(...arrayWithTiles);
```
Wystarczy `return arrayWithTiles`

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+    generateArrayWithUpdatedTiles(board) {
+        let arrayWithTiles = [];
+        for (let [index, subArray] of board.entries()) {
+            arrayWithTiles.push(subArray.filter(el => el).map(el => {
```
Nie lepiej utworzyć nowej metody, któa opisze co robić ten fragment kodu:
```
subArray.filter(el => el).map(el => {
                return {
                    id: el.id,
                    val: el.val,
                    y: index,
                    x: subArray.map(arrayEl => {
                        return arrayEl.id
                    }).indexOf(el.id),
                    mergedId : el.mergedId
                }
            })
```
?

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+                if (subArray[i].val === subArray[i + 1].val) return true;
+            }
+        }
+        tempBoard = this.rotateBoard([...tempBoard]);
```
Potrzebna tutaj jest ta desktrukturyzacja?

apyd replied:
Nieaktualne w związku ze zmianami

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    checkIfArraysEqual(arr1, arr2) {
+        if (typeof arr1 === 'undefined') return false;
```
A co jak `arr2` będzie undefined?

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    rotateBoard(boardToRotate) {
+        let rotatedBoard = Array(this.gameType).fill().map(() => Array());
```
Ta linijka wygląda też łudząco podobnie do tej z `initGame`

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+        }
+        return rotatedBoard;
+    }
+    mergeTiles(arr, key) {
```
Brak linijki odstępu

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+                    arr[i].mergedId = arr[i + 1].id;
+                    arr[i + 1] = '';
+                    this.updateScore(arr[i].val);
+                    controller.tilesOnBoard--;
```
Analogicznie tutaj - lepiej by klasa `Controller` miała metodę do bezpośredniej zmiany swojego pola

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+
+    checkIfPossibleMerge() {
+        let tempBoard = [...this.board];
+        for (const subArray of tempBoard) {
```
Czym jest `subArray`? W sensie to jest wiersz/kolumna na boardzie?

apyd replied:
Tak, to jest wiersz na boardzie, ale zmieniłem nazwę ponieważ ta była mylna.

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+                    arr[i].mergedId = arr[i - 1].id;
+                    arr[i - 1] = '';
+                    this.updateScore(arr[i].val);
+                    controller.tilesOnBoard--;
```
Analogicznie tutaj - lepiej by klasa `Controller` miała metodę do bezpośredniej zmiany swojego pola

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+    view
+} from './app.js';
+
+export default class Model {
```
Po przejrzeniu tego pliku zastanawiam się, czy najlepiej nie byłoby go podzielić na klasę `ScoreCounter` czy coś w tym stylu, który będzie odpowiedzialny za zliczanie punktów, aktualizowanie najlepszego wyniku, pobieranie najlepszego wyniku, zliczania ilości ruchów oraz na klasę `Tile` lub może `Board`, która by wykonywała operacje na boardzie

Mikailus replied:
PS. Wrócę do tego pliku przy następnym review, by lepiej zrozumieć transformacje w kodzie

Mikailus commented on assets/js/view.js:
```diff
@@ -0,0 +1,154 @@
+export default class View {
+    constructor() {
+        this.tileDimension;
```
Jak nie inicjalizujesz tych zmiennych w konstruktorze, to możesz je podać poza konstruktorem:
```
class MyClass {
  tileDimension;
  innerBorderWidth;
}
```

Mikailus commented on assets/js/view.js:
```diff
@@ -0,0 +1,154 @@
+        this.gameType;
+        this.screenType;
+        this.gameSettings = {
+            t4: {
```
1. Czym jest t4? Czym jest t5? Te nazwy niewiele mówią
2. Te wartości są wartościami początkowymi, inicjalizacyjnymi - można je wynieść do oddzielnego pliku, np. `game-settings.js`

```
export const gameSettings = {
...
}
```

i zaimportować tutaj

Mikailus commented on assets/js/view.js:
```diff
@@ -0,0 +1,154 @@
+    togglePopup() {
+        document.querySelector(".overlay").classList.toggle('hidden');
+    }
+    showEndGamePopup(time, numberOfMoves, result) {
```
Brak linijki odstępu

Mikailus commented on assets/js/view.js:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    clearBoard() {
+        let board = document.querySelector('.board');
```
const

Mikailus commented on assets/js/view.js:
```diff
@@ -0,0 +1,154 @@
+        }
+    }
+
+    initGame(gameType) {
```
`initializeGameView`

Mikailus commented on assets/js/view.js:
```diff
@@ -0,0 +1,154 @@
+        document.querySelectorAll('.scoreboard__value')[1].innerHTML = localStorage.getItem(`bestScore${this.gameType}`);
+        document.querySelectorAll('.scoreboard__value')[0].innerHTML = 0;
+        let windowWidth = window.innerWidth;
+        windowWidth > 480 ? this.screenType = 'desktop' : this.screenType = 'mobile';
```
`this.screenType = windowWidth > 480 ?  'desktop'  :  'mobile';`

Mikailus commented on assets/js/model.js:
```diff
@@ -0,0 +1,171 @@
+import {
+    controller,
+    model,
```
Czy plik model importuje sam siebie? :D

Mikailus commented on assets/js/view.js:
```diff
@@ -0,0 +1,154 @@
+        document.querySelector('.main').classList.add('hidden');
+    }
+
+    addTile({
```
Dziwne formatowanie

Mikailus commented on assets/js/view.js:
```diff
@@ -0,0 +1,154 @@
+            id
+        }) {
+        let board = document.querySelector('.board');
+        let tile = document.createElement('div');
```
Najlepiej utworzyć oddzielną metodę do tworzenia kafelka, np. `createTile` , która zwróci gotowy element do dołączenia do boarda

Mikailus commented on assets/js/controller.js:
```diff
@@ -0,0 +1,93 @@
+        view.showEndGamePopup(this.convertMillisToMinutesAndSeconds(timeElapsed), model.numberOfMoves, model.score);
+        this.gameStatus = 0;
+    }
+    convertMillisToMinutesAndSeconds(millis) {
```
Brak linijki odstępu

Mikailus commented on assets/js/controller.js:
```diff
@@ -0,0 +1,93 @@
+        this.gameStatus = 0;
+    }
+    convertMillisToMinutesAndSeconds(millis) {
+        var minutes = Math.floor(millis / 60000);
```
const

Mikailus commented on assets/js/controller.js:
```diff
@@ -0,0 +1,93 @@
+    }
+    convertMillisToMinutesAndSeconds(millis) {
+        var minutes = Math.floor(millis / 60000);
+        var seconds = ((millis % 60000) / 1000).toFixed(0);
```
const

Mikailus commented on assets/js/controller.js:
```diff
@@ -0,0 +1,93 @@
+
+    endGame() {
+        this.endGameTime = performance.now();
+        let timeElapsed = this.endGameTime - this.startGameTime;
```
const

Mikailus commented on assets/js/controller.js:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    stopGame() {
+        this.gameStatus = 0;
```
Najlepiej stworzyć plik `game-statuses.js`, który będzie wyglądałtak
```
export const gameStatuses {
  stopped: 0,
  costam: 1,
  costam2: 2
}
```
i posługiwać się w kodzie konkretnymi nazwami, np. `this.gameStatus = gameStatuses.stopped`

Mikailus commented on assets/js/controller.js:
```diff
@@ -0,0 +1,93 @@
+        if (type === 'touchend') {
+            let key;
+            if (Math.abs(Math.round(eY - this.initY)) < this.minSwipeDistance && Math.abs(Math.round(eX - this.initX)) < this.minSwipeDistance) return;
+            let angle = (Math.atan2(Math.round(eY - this.initY), (Math.round(eX - this.initX))));
```
```
calculateAngle(eX, eY) {
            let angle = (Math.atan2(Math.round(eY - this.initY), (Math.round(eX - this.initX))));
            angle *= 180 / Math.PI;
            return Math.round(angle);
}
```

Mikailus commented on assets/js/controller.js:
```diff
@@ -0,0 +1,93 @@
+            view.moveTiles(model.moveTiles(key));
+            model.canAddTile ? view.addTile(model.addTile()) : null;
+            if (this.tilesOnBoard === (this.gameType * this.gameType)) {
+                let isContinued = model.checkIfPossibleMerge();
```
const

Mikailus commented on assets/js/controller.js:
```diff
@@ -0,0 +1,93 @@
+            if (angle === 0) return;
+            else {
+                if (angle >= -135 && angle < -45) {
+                    key = 'ArrowUp';
```
Najlepiej stworzyć plik keys.js, który będzie wyglądał tak
```
export const keys {
  ArrowUp: 'ArrowUp',
  ...
}
```
i posługiwać się w kodzie konkretnymi nazwami, np. `key = keys.ArrowUp`

Mikailus reviewed (commented):
Jeszcze został index.html i view.js

apyd committed:
Mixin for generating tile color modifiers and function for color return

apyd committed:
Changes in styles to use new mixins/function

apyd committed:
Fix for last change (missing colors/autoprefixes), fixed functon to return color

apyd committed:
Fixes for last changes (mixin/function), changed footer link color

- Updated tileColors names
- Updated mixin to return background-color instead of color
- Updated function to return color as string if not found in colors

apyd committed:
Basic gulp setup to compile scss and run browsersync + gitignore for node modules

apyd committed:
Added autoprefixer to gulp config

apyd committed:
Replaced let/var with const where possible, changed functions name

apyd committed:
Removed classes for html/body

apyd committed:
CSS properties grouped by type, removed class for html and body

apyd committed:
Using gulp-dart-sass instead of gulp-sass due missing support for @use

apyd committed:
SCSS refactor

- replaced @import with @use
- changes to breakpoints values to handle correctly whole media query
- updated style.scss with namespaces for mixins/variables usage

apyd committed:
Changes in project structure

- Style.scss renamed to main.scss
- Created partials based on components in application
- Divided style.scss between component partials
- Updated gulp.scss to reflect changes
- Generating css in separated folder 'css'
- Created new partial for functions to separate it from mixins

apyd committed:
SCSS refactor (layout and popup)

apyd committed:
Class name update

apyd committed:
Class names update

apyd committed:
SCSS refactor (scoreboard, popup)

apyd committed:
SCSS refactor (button)

apyd committed:
SCSS refactor

- Replaced scoreboard with score component
- Replaced radio with game-option component
- Updated class names for elements
- Excluded overlay and scoreboard from components and moved to layout

apyd committed:
Game option component and removed scoreboard

apyd committed:
SCSS refactor

apyd committed:
Elements class name changes

apyd committed:
Elements class name changes

apyd committed:
SCSS refactor (layout, button)

apyd committed:
SCSS refactor (button, layout, popup)

apyd committed:
View - updated class names reference

apyd committed:
Elements class names change

apyd committed:
Updated class name references

apyd committed:
SCSS refactor (game-option)

apyd committed:
SCSS refactor (layout, game-option)

apyd committed:
SCSS refactor (board)

apyd committed:
Changed files indentation from spaces to tabs

apyd committed:
Changes within classes constructor content

- Left only initialized variables within constructor
- Declared variables moved before constructor
- gameSettings renamed to elementsDimensions and moved to separated file

apyd committed:
new createBoard method

apyd committed:
Code refactor

apyd committed:
Renamed 'addTile' to 'addTileToBoard', created new method createTile

apyd committed:
Model refactor

- Added method create tile,
- Tile coordinates are generated by external function from utils
- coordinates are stored as objecrt

apyd committed:
New controller method  - incrementNumberOfTilesOnBoard

apyd committed:
Model refactor

apyd committed:
Model and controller refactor

apyd committed:
Eslint/Prettier/Babel configuration

apyd committed:
ESLint rules update

apyd committed:
ESLint rules update

apyd committed:
Controller refactor

apyd committed:
ESLint rule update

apyd committed:
View refactor

apyd committed:
View and Controller refactor

apyd committed:
ESlint config update

apyd committed:
Model refactor

apyd committed:
App.js refactor

apyd committed:
View refactor

apyd committed:
Changes in game status

apyd committed:
Removed unused event passed on click

apyd committed:
Fixes for utils functions

Fixed:
- unable to move tiles on swipe
- adding tile even if no move on board detected

apyd committed:
RWD fix

apyd committed:
Mobile-first approach (SCSS refactor)

apyd committed:
Minor fixes

- reduced tile value length from 4 to 3 as condition to reduce font-size
- changing canAddTile flag on game initialization
- updated gameStatus on game end

apyd committed:
Model, view, controller refactor

apyd committed:
RWD fix

apyd committed:
Renamed images name to keep consistency

apyd committed:
Mixin to generate image paths for RWD

apyd committed:
Accessibility changes

Supporting keyboard usage to navigate through game

apyd committed:
Fix for touch screens

apyd committed:
Accessibility changes

apyd committed:
Accessibility changes

apyd committed:
Code refactor

- Model divided between score, tile and board components
- Removed circular dependencies
- Simplified utils functions

apyd committed:
Swipe gestures fix

apyd committed:
Fix - tiles size/position was not recalculated on screen resize

apyd committed:
Fix for showing popup on end game

apyd committed:
Fix for showing popup on end game

apyd committed:
Update README.md
